lib/shared/src/llm-providers/groq/chat-client.ts

```diff
@@ -0,0 +1,116 @@
+import { getCompletionsModelConfig } from '../..'
+import { onAbort } from '../../common/abortController'
+import { CompletionStopReason } from '../../inferenceClient/misc'
+import type { CompletionLogger } from '../../sourcegraph-api/completions/client'
+import type {
+    CompletionCallbacks,
+    CompletionParameters,
+    CompletionResponse,
+} from '../../sourcegraph-api/completions/types'
+
+const GROQ_CHAT_API_URL = new URL('https://api.groq.com/openai/v1/chat/completions')
+
+/**
+ * NOTE: Behind `chat.dev.models` configuration flag for internal dev testing purpose only!
+ *
+ * Calls the Gork API for chat completions.
+ * This also works with the OpenAI API or any OpenAI compatible providers.
+ * The endpoint can be changed via the apiEndpoint field in the `chat.dev.models` configuration.
+ */
+export function groqChatClient(
+    params: CompletionParameters,
+    cb: CompletionCallbacks,
+    completionsEndpoint: string,
+    logger?: CompletionLogger,
+    signal?: AbortSignal
+): void {
+    const log = logger?.startCompletion(params, completionsEndpoint)
+    if (!params.model || !params.messages) {
+        log?.onError('No model or messages')
+        throw new Error('No model or messages')
+    }
+
+    const config = getCompletionsModelConfig(params.model)
+    if (!config?.model) {
+        cb.onError(new Error(`Unknown model ${params.model}`))
+        return
+    }
+
+    const chatParams = {
+        model: config?.model,
+        messages: params.messages.map(msg => {
+            return {
+                role: msg.speaker === 'human' ? 'user' : 'assistant',
+                content: msg.text ?? '',
+            }
+        }),
+        stream: true,
+    }
+
+    fetch(config?.endpoint ?? GROQ_CHAT_API_URL, {
+        method: 'POST',
+        body: JSON.stringify(chatParams),
+        headers: {
+            'Content-Type': 'application/json',
+            Authorization: `Bearer ${config.key}`,
+        },
+        signal,
+    })
+        .then(res => res.body?.getReader())
+        .then(async reader => {
+            if (!reader) {
+                log?.onError('No response body')
+                throw new Error('No response body')
+            }
+
+            onAbort(signal, () => reader.cancel())
+
+            let responseText = ''
+            const textDecoder = new TextDecoder()
+
+            // Handles the response stream to accumulate the full completion text.
+            while (true) {
+                const { done, value } = await reader.read()
+                if (done) {
+                    cb.onChange(responseText)
+                    cb.onComplete()
+                    break
+                }
+
+                const chunk = textDecoder.decode(value, { stream: true })
+                const lines = chunk.split('\n')
+
+                for (const line of lines) {
+                    if (line?.startsWith('data: ')) {
+                        const data = line.slice(6)
```

Comment:
It's nice to explain where these magic numbers come from with variables.

```suggestion
                    const dataMarker = 'data: '
                    if (line?.startsWith(dataMarker)) {
                        const data = line.slice(dataMarker.length)
```